### ISSUE TITLE
Support all varieties of Null handling

### ISSUE THREAD

**Issue by Randgalt** [Randgalt/record-builder#111]
Yes that will probably be good for all. However I can't keep working on #106 or the other nullable annotation bugs till some decisions are made. Which is why I have been so back and forth on this as there are some serious previous design decisions that are impacting me.

So I'm fine with holding off indefinitely till some decisions are made (like lets copy what immutables does).

And I'm sorry for coming off rough and flooding the inbox but I have already sunken some time on this (a lot more than just changing some call from `of` to `ofNullable`) as I sort of felt obligated given previous discussions on reddit and whatnot.

I'll stop work and commenting till the dust clears.

_Originally posted by @agentgt in https://github.com/Randgalt/record-builder/issues/107#issuecomment-1104038505_

**Comment by Randgalt** [comment]
I'm opening this new issue to deal with any Null issues in Record-Builder. I'd like to stipulate a few things to keep in mind for this discussion and any solutions:

- Record-Builder is a targeted at a wide audience. That audience will contain users that are mostly concerned with utility over correctness or compliance with a spec
- Record-Builder has a well defined options/customization facility that can be used to generate builders and ancillary classes that are very specialized

Given this - what what needs to be done here. I disagree that the current implementation is wrong in any way. We can use the options to generate classes as needed. So, what should the generation be for the purposes of this issue? 

cc @agentgt

**Comment by agentgt** [comment]
See my comment here: https://github.com/Randgalt/record-builder/issues/106#issuecomment-1120409923

**Comment by Randgalt** [comment]
Thanks for the description - I'll see what can be done

**Comment by agentgt** [comment]
I forgot to add that there are three "use cases" of null handling you should consider:

 1. Compile time Static Analysis (what #106 was originally about)
 1. Runtime Assertion (requireNonNull) aka defensive programming
 1. Runtime Validation (Java Bean Validation API aka javax.constraints)

I know you talked about not really complying to a spec but about utility and the above are three use cases when dealing with null.

As for:

>  I disagree that the current implementation is wrong in any way. 

When I said the library was "wrong" it was because you or someone who did the `NotNull` for "Full Record support" made number 3... aka Validation do all three of those by accident. This was especially bad because static analysis and runtime assertion can be a direct contradiction of what `javax.constraint.validation` api needs. 

That is a very common use case is to have:

```java
@checkerframework.Nullable @javax.constraint.validation.NotNull String someField; 
```

(replace checkerframework  with any static analysis annotation)

However the generated code does a:

```
withSomeField(@javax.constraint.validation.NotNull /* <- that validation annotation should not be there and the nullable one is missing because its TYPE_USE */ String someField) {
Objects.requireNonNull(someField); // and this is wrong because we expect it to be null
}
```

How can you ever validate an object if you cannot make it because NPE is thrown? That is `@javax.constraint.validation.NotNull` does not mean never null... it means the opposite. It means we expect it to be null and we will tell the user.  Furthermore I believe `javax.constraint.validation` in some cases doesn't even want the annotations propagated (in fact it used to fail if you annotated a getter and a field)

The irony is if you do use static analysis (you need far less to almost zero defensive programming with static analysis) and java validation the only time a `requireNonNull` would ever be needed is if you wanted to support `Optional` (which is exactly what immutables does).


I explained this extensively over and over again in #107 . See this comment validation api: https://github.com/Randgalt/record-builder/issues/107#issuecomment-1102624501

So it's not "wrong" in terms of it doesn't crash. In fact it arguable other than not supporting TYPE_USE did indeed propagate the annotation. However it certainly doesn't do out of the box what most folks who know what the validation api is supposed to do. It certainly doesn't do what Spring expects in its `@Valid` support (because you would blow up before the object made it to the controller). This all because the default regex on `NotNull` annotations. (However don't forget you did show the example with javax.validation w/ your gist).

So since you don't want to comply to spec I recommend the best solution is to do what Immutables does and make your own annotations for the various null handling. And while Immutables isn't your goal in my opinion it currently does a better job than this library because the exact scenario I had with validation/static analysis contradiction didn't happen on my first stab at the library (it also has semver... not that I love semver it does give me some idea of the backward compatibility). 

It is not about being "wrong" or "right". I should not have said that. That wasn't my point. It was about making people use the other libraries you plan on making "specialized" classes for correctly as those do have specs and designs. It's about being aware of a specific portion of annotations that has been around since Java 8. Static analysis loves code generation (since reflection cannot be verified) and annotation processing. Its about knowing validation needs nulls. It's also about planning things correctly from the beginning so you don't break peoples code later on every upgrade or make a trillion little switches of configuration. 

You advertise the library frequently on reddit so imagine I won't be the only once with some opinion that what the libraries is not exactly the "right" way. 

So it is not "wrong" but can we agree it needs to be improved?

**Comment by Randgalt** [comment]
As I said, I've never encountered a TYPE_USE annotation and thus don't know what to do. I'm happy to it make it work properly but I need to understand what is wrong and what can be improved. I hope we can just focus on the problem and eliminate any superfluous comments. For example, in #106 you say:

> Let us assume `@Nullable` is a TYPE_USE like it is for JSpecify.

```java

    @RecordBuilder
    public record MyRecord(@Nullable String name, String required) {}

```

> The record builder should copy the type `java.lang. @Nullable String` for wherever `name` is used. 

I searched around and the docs for TYPE_USE (even that JDK bug report) are opaque. Wouldn't it also be correct to just copy the TYPE_USE `@Nullable` without the FQPN? e.g.

```
// in builder...
public void name(@Nullable String name) {
   ...
}
```

My read of the spec says that the `@Nullable` would bind to the type `String`. Record-Builder doesn't use FQPNs. 

It's a 2 line code change to get the Record Builder processor to recognize the TYPE_USE annotations. Let me know if this is correct for `numbers`:

```
@RecordBuilderFull
public record FullRecord(@TypeUseNullable @NotNull List<Number> numbers, @NotNull Map<Number, FullRecord> fullRecords, @NotNull String justAString) {
}
```

```
    @Generated("io.soabase.recordbuilder.core.RecordBuilder")
    public FullRecordBuilder numbers(
            @TypeUseNullable @NotNull Collection<? extends Number> numbers) {
        this.numbers = (numbers != null) ? new ArrayList<>(numbers) : null;
        return this;
    }

    /**
     * Return the current value for the {@code numbers} record component in the builder
     */
    @Generated("io.soabase.recordbuilder.core.RecordBuilder")
    @TypeUseNullable
    @NotNull
    public List<Number> numbers() {
        return numbers;
    }
```

**Comment by agentgt** [comment]
What happens when you have name conflicts? A dumb example is `com.stuff.String` and `java.lang.String`. At some point I assume you will need to print the FQN?

No it will not work for all cases. You also need to worry static inner classes, arrays and generics.

For example you can have a record like:

```java
record MyRecord(
    Map<String, @Nullable String> model, 
    @NonNull String @Nullable [] ) {
}
```

You basically need to print the type exactly as it is and that is what the JDK bug is about. You should be able to just call `toString` on the TypeMirror so you do not have to use a crazy complicated Visitor in the annotation processor to recreate it (as well as deal with name conflicts).

In fact you probably do not even need to try to interpret the meaning of the TYPE_USE annotations but just propagate them. That is all you have to do.

However your solution might just be good enough (e.g. 80-20) till the JDK makes it easier.


----

Apologies on the superfluous comments but it is very complicated subject and is extremely frustrating when someone is dismissive with "I am not wrong" or the implementation "is not wrong". That makes me feel like I have to go out of my way to show its problems.

**Comment by Randgalt** [comment]
FYI - I did a lot of playing around/investigating. In order to get a complete solution to the various possible positions of `TYPE_USE` annotations I'll need to make some changes to JavaPoet. JavaPoet appears to be abandoned by Square so I could fork it. I did some testing with a fork and was able to solve most of the issues. However, I feel that this is a step too far. We can get an 80% solution that solves how most people use Checker framework's null annotation with a few fixes in Record-Builder. I'll have a PR with that soon. Afterwards, we can consider forking JavaPoet or some other alternative.

**Comment by brainbytes42** [comment]
Hi, 
really cool tool - and maybe it's already possible, but I didn't find it: How can I make may complete record null-safe without annotation each and every single parameter or using other frameworks?
I searched for something like interpretNotNulls in RecordBuilder.Options, in my case maybe "requireNonNull" or even better "defaultRequireNonNull" to allow for some explicitly annotated nullable values...
(If I need to annotate each and every field, this might be errorprone and repetitive; and for the default-non-null, it would be more obvious which values are nullable, other than when all parameters are annotated and look visually almost the same.)

**Comment by NicklasWallgren** [comment]
Are there any annotation similar to lomboks `NonNull` which adds `Objects.requireNonNull()` or similar?

**Comment by Randgalt** [comment]
> Are there any annotation similar to lomboks `NonNull` which adds `Objects.requireNonNull()` or similar?

I played around with some ideas but they all required bytecode modifications and I'd like to stay away from that. See here: https://github.com/Randgalt/record-builder/discussions/120

**Comment by CfFanDuel** [comment]
Quite an old issue to dig up but was there any ever progress on the `TYPE_USE` annotation issue?

I've come across it whilst trying to use JSpecify's `@Nullable` annotation with a staged builder rather than my own custom annotation and it doesn't work :( 

My custom annotation was propagated as I expected it to be and was hoping the JSpecify annotations would be too as it makes this library easily compatible with null checking frameworks.

**Comment by agentgt** [comment]
@CfFanDuel The issue is that the JDK has a bug in the annotation processor. `TYPE_USE` annotations are not visible across compile time boundaries.

https://github.com/jspecify/jspecify/issues/365

If all your code that has jspecify-like annotations is in one compile time project (e.g. Maven module) then you will not have a problem in terms of annotation propagation (discovery by an annotation processor).

If however that is not the case your other modules will appear to have no `@Nullable` annotations. I and many others have complained about this but it is not easy for the JDK to make this switch because of various unknown consequences. As in there it is unlikely this bug will be backported.

Some of the unknown consequences is just toString-ing TypeMirror or TypeElement for the FQN which tons of code does. I say that because IIRC the annotation would be attached to the type. So you can imagine how many annotation processors might break on this but that is technically a different bug.

The Eclipse JDT annotation processor I think already does this and I think it even works across compile time boundaries.  So old code that cannot handle the annotation need to do sort of like a type erasure and remove the annotation if they are doing string mangling of the type name.

Anyway the Spring team has far more clout than I and others and they are now complaining about it so maybe this will get fixed in JDK 25.

**Comment by msridhar** [comment]
> As for the JDK bug itself I believe there's a backport to Java 21 though you have to enable it via a compiler option I think. Not sure if it's the same in 25.

No flag needed for JDK 25 javac; the bug is just fixed.  For JDK 21 javac the flag is `-XDaddTypeAnnotationsToSymbol=true`.  Not sure I fully follow the discussion here, but if the only issue was the javac bug with type use annotations, then yes, it should just be fixed on JDK 25.

**Comment by CfFanDuel** [comment]
I'll try to test with JDK 25 at some point and then report back :)
